### PR TITLE
[fix] handle better axes going just at the limit

### PR DESCRIPTION
**Pull request by pieleric**

### CODE REVIEW

**Review by nandishjpatel — APPROVED** [review]

File: install/linux/usr/share/odemis/sim/sparc2-time-correlator-sim.odm.yaml [hunk 1]
```diff
@@ -164,7 +164,7 @@ SPARC2-VisNIR-TR: {
         address: null,
         axes: ["l1", "l2", "spec-sel", "fiby", "slit"],
         # These values only need to be roughly correct
-        ustepsize: [25.1e-9, 25.1e-9, 25.1e-9, 5.86e-9, 5.e-9], # m/µstep
+        ustepsize: [25.1e-6, 25.1e-6, 25.1e-6, 5.86e-6, 5.e-7], # m/µstep
```
Comment (nandishjpatel) [inline install/linux/usr/share/odemis/sim/sparc2-time-correlator-sim.odm.yaml:167]:
```suggestion
        ustepsize: [25.1e-6, 25.1e-6, 25.1e-6, 5.86e-6, 5.e-6], # m/µstep
```
Maybe?

File: install/linux/usr/share/odemis/sim/sparc2-time-correlator-sim.odm.yaml [hunk 2]
```diff
@@ -421,8 +412,9 @@ SPARC2-VisNIR-TR: {
     init: {
         axis_name: "od",
         # position when referenced (adjusted to be a multiple of the actuator stepsize)
-        offset: -3.00000925, # od
-        ref_start: 2.8 #  a little before the reference (inverted axis)
+        offset: -2.98, # od
+        ref_start: 2.78, #  a little before the reference (inverted axis)
+        ref_period: null,  # never automatically reference the axis
```
Comment (sneepstefan) [inline install/linux/usr/share/odemis/sim/sparc2-time-correlator-sim.odm.yaml:417]: Could you update the comment with the unit of the reference period when the auto reference is needed?

File: src/odemis/acq/stream/_base.py [hunk 3]
```diff
@@ -681,8 +681,13 @@ def _update_linked_position(self, act, pos):
...
             # update va
-            va.value = axpos
-            logging.info("Updating local axis %s to position %s", va_name, axpos)
+            axpos_clipped = va.clip(axpos)  # In case the axis went out of range
+            va.value = axpos_clipped
+            if axpos == axpos_clipped:
+                logging.info("Updating local axis %s to position %s", va_name, axpos)
+            else:
+                logging.warning("Updating local axis %s to clipped position %s (actual pos = %s)",
+                                va_name, axpos_clipped, axpos)
```
Comment (sneepstefan) [inline src/odemis/acq/stream/_base.py:690]: Shouldn't the message be the other way around? e.g. if the axis position is the same as the axis position clipped -> Updating local axis %s to clipped position. Or do I have to read it differently?

File: src/odemis/driver/tmcm.py [hunk 4]
```diff
@@ -417,7 +417,10 @@ def __init__(self, name, role, port, axes, ustepsize, address=None,
             if sw_rng is not None:
                 if not sw_rng[0] <= 0 <= sw_rng[1]:
                     raise ValueError("Range of axis %d doesn't include 0: %s" % (i, sw_rng))
-                phy_rng = (max(phy_rng[0], sw_rng[0]), min(phy_rng[1], sw_rng[1]))
+                # Make sure the range is a multiple of the ustepsize
```
Comment (sneepstefan) [inline src/odemis/driver/tmcm.py:420]: This sounds logical but why is that necessary? Probably due to issues in the field with the tmcm?

**Review by sneepstefan — CHANGES_REQUESTED** [review]

File: src/odemis/acq/stream/_base.py (hunk 3 above)
Comment (pieleric) [reply src/odemis/acq/stream/_base.py:690]: If the clipped value is the same are the original value, it means the "clipping" had no effect, which means it wasn't clipped. So we say it's been clipped if the clipped value is different from the original value. The code is correct.

File: src/odemis/driver/tmcm.py (hunk 4 above)
Comment (pieleric) [reply src/odemis/driver/tmcm.py:420]: Good catch. The comment needs to explain *why* (not *what*). Essentially, it's because if it's not rounded and the user moves to the limit, the hardware will round the value, and there is 50% chances it ends up a tiny bit outside the range... which confuses some code (and some users).

**Review by sneepstefan — APPROVED** [review]
